### PR TITLE
Fix alignment and Undefined Behavior

### PR DESCRIPTION
**Pull request by elichai**
Hi,
1. you cannot access T if the pointer isn't aligned to T. so the alignment needs to be to T, not to u8.

2. both `mem::replace` and `*inner = t` deference a zeroed memory, this is instant UB for every type that zero isn't a valid value (i.e. NonNull/NonZero etc.).

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -17,9 +17,8 @@ impl<T: Default> Trailer<T> {
     pub fn new(capacity: usize) -> Trailer<T> {
         unsafe {
             let trailer = Trailer::allocate(capacity);
-            let previous = mem::replace(&mut *(trailer.ptr as *mut T), T::default());
-            mem::forget(previous);
-
+            let ptr = trailer.ptr as *mut T;
+            ptr.write(T::default());
```
Comment (elichai) [inline src/lib.rs:21]: ptr write is the write way to write into an invalid/uninitialized part of memory

File: src/lib.rs [hunk 2]
```diff
@@ -29,8 +28,8 @@ impl<T: Copy> Trailer<T> {
     pub fn from(t: T, capacity: usize) -> Trailer<T> {
         unsafe {
             let trailer = Trailer::allocate(capacity);
-            let inner = trailer.ptr as *mut T;
-            *inner = t;
+            let ptr = trailer.ptr as *mut T;
+            ptr.write(t);
```
Comment (elichai) [inline src/lib.rs:32]: Same not about `ptr.write`

File: src/lib.rs [hunk 3]
```diff
@@ -40,7 +39,7 @@ impl<T: Copy> Trailer<T> {
 impl<T> Trailer<T> {
     unsafe fn allocate(capacity: usize) -> Trailer<T> {
         let size = mem::size_of::<T>() + capacity;
-        let align = mem::align_of::<u8>();
+        let align = mem::align_of::<T>();
```
Comment (elichai) [inline src/lib.rs:42]: This is required to make sure the pointer is aligned to `T`, otherwise you can't write `T` into it.

**Comment by Geal** [comment]
Thanks! sorry for the delay

**Pull request merged**